content/docs/specs/remote_write_spec_2_0.md

```diff
@@ -130,18 +130,42 @@ Senders MUST include a user agent header that SHOULD follow [the RFC 9110 User-A
 
 ### Response
 
-Receivers ingesting all samples successfully MUST return a 200 HTTP status code. In such a successful case, the response body from the Receiver SHOULD be empty; Senders MUST ignore the response body. The response body is RESERVED for future use.
+Receivers ingesting all Samples and Histograms successfully MUST return a [success 2xx HTTP status code](https://www.rfc-editor.org/rfc/rfc9110.html#name-successful-2xx). In such a successful case, the response body from the Receiver SHOULD be empty and the status code SHOULD be [204 HTTP No Content](https://www.rfc-editor.org/rfc/rfc9110.html#name-204-no-content); Senders MUST ignore the response body. The response body is RESERVED for future use.
```

Comment:
> In such a successful case, the response body from the Receiver SHOULD be empty and the status code SHOULD be [204 HTTP No Content](https://www.rfc-editor.org/rfc/rfc9110.html#name-204-no-content);

What's the reasoning for changing to 204?

Reply:
Great question! 

First of all this was previously wrong as we technically NOT allowed 204 ;p so I moved this to 2xx as in 1.0 https://prometheus.io/docs/specs/remote_write_spec/#:~:text=Prometheus%20remote%20Write%20compatible%20receivers%20MUST%20respond%20with%20a%20HTTP%202xx%20status%20code%20when%20the%20write%20is%20successful.

Secondly I noticed Prometheus handler and others use [204](https://www.rfc-editor.org/rfc/rfc9110.html#name-204-no-content), not 200. I dived deeper and the reason is simple - we won't write any response, so correct response is 204. This helps clients to be more efficient and assume no body (in theory).

Thanks for challenging, because now I might say we don't need this SHOULD. RFC 9110 states that no response means 204, so no need to specify this in our spec. However one benefit of specifying this here is that not many might now about 204, so worth highlighting? I think generally I would keep it 🙈 No harm in highlighting, it's just a SHOULD and takes a few words really. 🤔

Reply:
Actually since you are here I noticed another change we did in the past vs 1.0.

1.0 states:

> Prometheus remote Write compatible receivers MUST respond with a HTTP 2xx status code when the write is successful

We state:

> Receivers ingesting all Samples and Histograms successfully MUST return a [success 2xx HTTP status code](https://www.rfc-editor.org/rfc/rfc9110.html#name-successful-2xx).

In some way we specify **the success condition**. In 1.0 they were up to Receiver. More open definition allow variety of Receive use cases perhaps? Not sure yet if this good or bad, but it matters for Response Header Stats. Let's iterate with some examples (helps me too - thinking loudly):

A) Like a validation receiver that only checks stuff and NOT ingest (like literally our compliance test) - it might not ingest (as in-store anywhere) anything and still return true, so even "ingesting" is dubious here. 

B) What if Receiver does not support native histograms yet? If we kept open definition from 1.0 it could either return 400 on those samples or 200 (silent ignore). In current 2.0 it HAS to return 400 partial error (which is kind of nice, because receivers are more predictable form sender perspective)?

C) What if somebody creates a Receiver that ONLY wants native histograms but nothing else. Then from this Receiver perspective success means ingested histograms, but samples would cause 400 effectively blocking this kind of receiver from being implemented (likely no senders would be happy to have mode of "skip samples"). 

D) What if somebody creates a Receiver that ONLY wants exemplars but nothing else (like a separate store). This sucks a lot now because it will always have to return 400 kind of, so can't silently drop other stuff (maybe more predictable again?)

E) What if future Sender wants to ensure exemplars are part of success (ultra durability of everything). Sender can't rely on 2xx anymore but has to check Stats. I guess that's fine, with 1.0 undefined success they cannot technically rely on 2xx even unless they know implementation of Receiver.

So generally.. so far it feels to me, that binding the success definition to Remote Write and not Receiver implementation is helping because Sender can rely on spec more, less on different kinds of implementations 🤔  I remember we discussed a bit about same in the past, but can't find it now 🙈 

Now, for the response headers it matters, because of that assumption that if you got 2xx AND (accepted samples == 0 and/or accepted histograms == 0) AND you sent some samples/histograms == you talk to 1.0 receiver accidently. If success was not defined.. we couldn't assume so. Maybe there are simpler ways to achieve so? I ruled out `X-Remote-Write-Version` in Response HTTP header, because it would still not guarantee bug didn't happened in the content negotiation. @npazosmendez proved alternative of deterministic unmarshal fail is also out of question. 

Anyway, will add those reasoning to proposal for future readers. Wanted to rechallenge and confirm we are ok here (:

Reply:
And finally I defined `samples` in previous statement meant Samples and Histograms... but maybe we wanted to include Exemplars too? In some way "successful write" of Sample with Exemplar is only successful when exemplar is also ingested? Maybe not? Maybe it's fine to not specify fully and leave that to Receiver?

Reply:
We found some consensus in https://cloud-native.slack.com/archives/C06B8KJUU2Y/p1720628447889389 will try to put that on the PR.